polytope_server/common/observability/otel.py

```diff
@@ -0,0 +1,86 @@
+from opentelemetry import trace
+from opentelemetry.propagate import inject, extract
+from contextlib import contextmanager
+import logging
+
+def add_trace_context(request):
+    carrier = {}
+
+    # Inject the context into the carrier
+    inject(carrier)
+
+    # Adding trace_id and span_id to the request data for logging
+    request.otel_trace_ctx = {
+        'carrier': carrier  # Store the injected carrier
+    }
+
+    current_span = trace.get_current_span()
+    current_span.set_attribute("polytope.request.id", request.id)
+
+    logging.debug(f"[OTEL] Request created with carrier: {request.otel_trace_ctx['carrier']}")
+
+def update_trace_context(request):
+    carrier = {}
+
+    # Inject the context into the carrier
+    inject(carrier)
+
+    # Adding trace_id and span_id to the request data for logging
+    request.otel_trace_ctx = {
+        'carrier': carrier  # Store the injected carrier
+    }
+
+    logging.debug(f"[OTEL] Request created with carrier: {request.otel_trace_ctx['carrier']}")
+
+
+def restore_trace_context(request):
+    # If otel context is not set, return
+    if not hasattr(request, 'otel_trace_ctx'):
+        return
+
+    logging.debug(f"[OTEL] Restoring context from carrier: {request.otel_trace_ctx['carrier']}")
+
+    # Extract the context from the stored request's carrier
+    extracted_context = extract(carrier=request.otel_trace_ctx['carrier'])
+
+    current_span = trace.get_current_span()
+    current_span.set_attribute("polytope.request.id", request.id)
+
+    return extracted_context
+
+@contextmanager
+def create_new_span_internal(span_name, request_id=None, parent_context=None, kind=trace.SpanKind.INTERNAL):
+    with create_new_span(span_name, request_id, parent_context, kind) as span:
+        span.set_attribute("role", "internal")
+        yield span
+
+# Forcing span kind Server because of AWS representation
+@contextmanager
+def create_new_span_producer(span_name, request_id=None, parent_context=None, kind=trace.SpanKind.SERVER):
+    with create_new_span(span_name, request_id, parent_context, kind) as span:
+        span.set_attribute("role", "producer")
```

Comment:
Is there a reason to do this as four separate methods rather than passing the role as an argument? The only one that is odd is `create_new_span_internal` since there is only one valid mapping from kind to role.

In the case of passing the role, it could be passed as an enum value similarly to `kind` so that we limit what the options are to these three methods

Reply:
yes you are right, during testing, I frequently had to switch between Consumer, Producer, and Server span kinds. To simplify this and avoid modifying each service individually, I created dedicated functions for each type. This way, it's clear that, for example, a Producer span is being forced to a Server kind due to AWS requirements. Following that logic, it felt inconsistent to leave Internal spans handled differently, so I aligned it with the same approach.